### PR TITLE
feat(vm): add metrics

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Add new metrics 
```
# VM core count
d8_virtualization_virtualmachine_configuration_cpu_cores = 1 ... 1024
# VM coreFraction param without % 
d8_virtualization_virtualmachine_configuration_cpu_core_fraction = 5 10 25 50 100

# VM requested cores like kube_pod_container_resource_requests{resource="cpu"}
d8_virtualization_virtualmachine_configuration_cpu_requested_cores = 0.1 0.25 1 10 100 ...
# VM system overhead (vmpod cores - vm cores) like kube_pod_container_resource_requests{resource="cpu"}
d8_virtualization_virtualmachine_configuration_cpu_runtime_overhead = 0.1 0.25 1 10 100 ...

# VM memory size in bytes like kube_pod_container_resource_requests{resource="memory"}
d8_virtualization_virtualmachine_configuration_memory_size = bytes
# VM memory overhead in bytes (vmpod mem - vm mem) like kube_pod_container_resource_requests{resource="memory"}
d8_virtualization_virtualmachine_configuration_memory_runtime_overhead = bytes

# AwaitingRestartToApplyConfiguration condition (true | false)
d8_virtualization_virtualmachine_awaiting_restart_to_apply_configuration = 0 | 1
# ConfigurationApplied condition (true | false)
d8_virtualization_virtualmachine_configuration_applied = 0 | 1
# .spec.runPolicy (true | false)
d8_virtualization_virtualmachine_configuration_run_policy{run_policy="AlwaysOn|..."} = 0|1
# VM Pod name (0 - inactive | 1 - active)
d8_virtualization_virtualmachine_pod{pod_name="current-pod-name"} = 0 | 1
```

Base labels
` {name="vmname" namespace="ns" node="vmnode" uid="uid"}`

## Why do we need it, and what problem does it solve?
we need more metrics

## Checklist
- [X] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: api/core/v1alpha2/virtual_machine.go [hunk 1]
```diff
@@ -159,58 +233,79 @@ const (
...
+	// The name of the node on which the VM is currently running.
+	Node string `json:"nodeName"`
+	// Name of `virtualMachineIPAddressName` holding the ip address of the VirtualMachine.
+	VirtualMachineIPAddress string `json:"virtualMachineIPAddressName"`
+	// IP address of VM.
+	IPAddress string `json:"ipAddress"`
+	// The list of attached block device attachments.
+	BlockDeviceRefs []BlockDeviceStatusRef                   `json:"blockDeviceRefs,omitempty"`
+	GuestOSInfo     virtv1.VirtualMachineInstanceGuestOSInfo `json:"guestOSInfo,omitempty"`
+	// Specific points in VM's runtime.
```
Comment (diafour) [inline api/core/v1alpha2/virtual_machine.go:246]:
```suggestion
	// Detailed state of the virtual machine lifecycle.
```

File: api/core/v1alpha2/virtual_machine.go [hunk 2]
```diff
@@ -159,58 +233,79 @@ const (
...
-	//	currentValue any (bool|int|string|struct|array of structs)
-	//	desiredValue any (bool|int|string|struct|array of structs)
-	//
+	// * operation enum(add|remove|replace)
+	// * path string
+	// * currentValue any (bool|int|string|struct|array of structs)
+	// * desiredValue any (bool|int|string|struct|array of structs)
 	// Such 'any' type can't be described using the OpenAPI v3 schema.
 	// The workaround is to declare a whole change operation structure
 	// using 'type: object' and 'x-kubernetes-preserve-fields: true'.
```
Comment (diafour) [inline api/core/v1alpha2/virtual_machine.go:262]: This comment should not be a part of CRD. It is internal information for developers, not for users.

File: crds/doc-ru-virtualmachineclasses.yaml [hunk 3]
```diff
@@ -51,7 +51,6 @@ spec:
                                       `NotIn`  - значение ключа лейбла не входит в заданный список лейблов `values`.
                                       `Exist` - ключ лейбла существует.
                                       `DoesNotExist`  - ключа лейблане существует.
```
Comment (diafour) [inline crds/doc-ru-virtualmachineclasses.yaml:53]:
```suggestion
                                      `DoesNotExist`  - ключа лейбла не существует.
```

File: crds/doc-ru-virtualmachines.yaml [hunk 4]
```diff
@@ -649,3 +649,33 @@ spec:
...
+                    properties:
+                      active:
+                        description: Отметка активного пода.
+                      name:
+                        description: Имя пода.
+                resources:
+                  description: Статистика использования ресурсов.
+                  properties:
+                    cpu:
+                      description: Статистика использования ресурсов ЦП.
```
Comment (diafour) [inline crds/doc-ru-virtualmachines.yaml:665]:
```suggestion
                      description: Статистика использования ресурсов ЦПУ.
```

File: crds/virtualmachineclasses.yaml [hunk 5]
```diff
@@ -12,6 +12,7 @@ spec:
   group: virtualization.deckhouse.io
   names:
     categories:
+      - all
```
Comment (diafour) [inline crds/virtualmachineclasses.yaml:15]: Please, no. Cluster wide resource should not has 'all' category.

File: images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/collector.go [hunk 6]
```diff
@@ -17,40 +17,135 @@ limitations under the License.
...
+	),
+	MetricVirtualMachinePod: prometheus.NewDesc(prometheus.BuildFQName(metrics.MetricNamespace, "", MetricVirtualMachinePod),
+		"The virtualmachine current active pod.",
+		WithBaseLabels("pod"),
+		nil,
+	),
 }
 
-func SetupCollector(vmLister Lister, registerer prometheus.Registerer) *Collector {
+func SetupCollector(reader client.Reader, registerer prometheus.Registerer) *Collector {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/collector.go:132]: Consider split this file into 3: collector.go with Collector, metrics.go with metrics definitions and a scraper.go with the scraper.

**Comment by diafour** [comment]
Let's rename virtualmachine.yaml into virtualmachines.yaml in a separate PR and merge it to main to see difference between hand-made and generated CRD in this PR.

File: crds/virtualmachines.yaml [hunk 7]
```diff
@@ -0,0 +1,1660 @@
...
+                            description: |-
+                              matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+                              map is equivalent to an element of matchExpressions, whose key field is "key", the
+                              operator is "In", and the values array contains only "value". The requirements are ANDed.
+                            type: object
+                        type: object
+                        x-kubernetes-map-type: atomic
+                      matchLabelKeys:
+                        description: |-
+                          MatchLabelKeys is a set of pod label keys to select the pods over which
```
Comment (fl64) [inline crds/virtualmachines.yaml:1133]: Pods need to be replaced by virtual machines here and throughout the CRD specification

File: crds/virtualmachines.yaml [hunk 8]
```diff
@@ -0,0 +1,1660 @@
...
+                      - size
+                    type: object
+                  type: array
+                conditions:
+                  description: Specific points in VM's runtime.
+                  items:
+                    description:
+                      "Condition contains details for one aspect of the current
+                      state of this API Resource.\n---\nThis struct is intended for
+                      direct use as an array at the field path .status.conditions.  For
```
Comment (fl64) [inline crds/virtualmachines.yaml:1331]: do we really need this here?

File: crds/virtualmachines.yaml [hunk 9]
```diff
@@ -0,0 +1,1660 @@
...
+          name: IPAddress
+          type: string
+        - description: Time of creation resource.
+          jsonPath: .metadata.creationTimestamp
+          name: Age
+          type: date
+      name: v1alpha2
+      schema:
+        openAPIV3Schema:
+          description: VirtualMachine specifies configuration of the virtual machine.
```
Comment (fl64) [inline crds/virtualmachines.yaml:66]: you missed this
https://github.com/deckhouse/virtualization/blob/main/crds/virtualmachine.yaml#L30-L36

please synchronize the document with the original

File: crds/virtualmachines.yaml [hunk 10]
```diff
@@ -0,0 +1,1660 @@
...
+    kind: VirtualMachine
+    listKind: VirtualMachineList
+    plural: virtualmachines
+    shortNames:
+      - vm
+      - vms
+    singular: virtualmachine
+  scope: Namespaced
+  versions:
+    - additionalPrinterColumns:
```
Comment (fl64) [inline crds/virtualmachines.yaml:26]: please add
```
        - name: Migratable
          type: string
          jsonPath: .status.conditions[?(@.type=="Migratable")].status
          priority: 1
```

and other priorities and order like in origin

**Review by fl64 — CHANGES_REQUESTED** [review]

File: crds/virtualmachines.yaml (hunk 8 above)
Comment (yaroslavborbat) [reply crds/virtualmachines.yaml:1331]: It was auto generated by kubebuilder. The imported type includes this description. Should I exclude this?

File: images/virtualization-artifact/pkg/controller/vm/internal/statistic.go [hunk 11]
```diff
@@ -52,15 +55,144 @@ func (h *StatisticHandler) Handle(ctx context.Context, s state.VirtualMachineSta
...
 		return reconcile.Result{}, err
 	}
-
 	h.syncStats(current, changed, kvvmi)
+
+	pods, err := s.Pods(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pod, err := s.Pod(ctx)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/statistic.go:64]: What is the difference between pod and pods?

File: images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/unsafe.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	return &iterator{
+		reader: reader,
+	}
+}
+
+type iterator struct {
+	reader client.Reader
+}
+
+func (l *iterator) Iter(ctx context.Context, h handler) error {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/unsafe.go:37]: Not public if struct is private

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/statistic.go (hunk 11 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vm/internal/statistic.go:64]: Pods - all pods of vm
Pod - active pod

**Review by diafour — APPROVED** [review]

File: images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/scrapper.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package virtualmachine
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog/v2"
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/monitoring/metrics/virtualmachine/scrapper.go:21]: How about replacing klog with slog in the new files? Slog is our go-to solution.

**Review by Isteb4k — DISMISSED** [review]

File: crds/virtualmachines.yaml [hunk 14]
```diff
@@ -1137,42 +1137,121 @@ spec:
...
+                            - type: integer
+                            - type: string
+                          description: Requested cores.
+                          pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
+                          x-kubernetes-int-or-string: true
+                        runtimeOverhead:
+                          anyOf:
+                            - type: integer
+                            - type: string
+                          description: runtime overhead.
```
Comment (fl64) [inline crds/virtualmachines.yaml:1180]:
```suggestion
                          description: Runtime overhead.
```

File: crds/virtualmachines.yaml [hunk 15]
```diff
@@ -1137,42 +1137,121 @@ spec:
...
+                          pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
+                          x-kubernetes-int-or-string: true
+                      required:
+                        - cores
+                      type: object
+                    memory:
+                      description:
+                        MemoryStatus defines statistics about the Memory
```
Comment (fl64) [inline crds/virtualmachines.yaml:1188]:
```suggestion
                        Memory utilization statistics.
```

File: crds/virtualmachines.yaml [hunk 16]
```diff
@@ -1137,42 +1137,121 @@ spec:
...
+                      name:
+                        description: Name of virtual machine pod.
+                        type: string
+                    required:
+                      - active
+                      - name
+                    type: object
+                  type: array
+                resources:
+                  description: ResourcesStatus defines resource usage statistics.
```
Comment (fl64) [inline crds/virtualmachines.yaml:1156]:
```suggestion
                  description: Resource utilization statistics.
```

File: crds/virtualmachines.yaml [hunk 17]
```diff
@@ -1137,42 +1137,121 @@ spec:
...
+                  properties:
+                    cpu:
+                      description:
+                        CPUStatus defines statistics about the CPU resource
```
Comment (fl64) [inline crds/virtualmachines.yaml:1160]:
```suggestion
                        CPU utilization statistics.
```

File: crds/virtualmachines.yaml [hunk 18]
```diff
@@ -1137,42 +1137,121 @@ spec:
...
+                        usage.
+                      properties:
+                        coreFraction:
+                          description: Current CoreFraction.
+                          type: string
+                        cores:
+                          description: Current number of cores inside the VM.
```
Comment (fl64) [inline crds/virtualmachines.yaml:1167]:
```suggestion
                          description: The current number of cores in the virtual machine.
```

**Review by fl64 — CHANGES_REQUESTED** [review]

**Review by fl64 — APPROVED** [review]

**Pull request merged**